File: client/src/main/java/dev/hoot/api/items/Trade.java
package dev.hoot.api.items;

import dev.hoot.api.commons.Predicates;
import dev.hoot.api.game.Game;
import dev.hoot.api.game.Vars;
import dev.hoot.api.widgets.Dialog;
import dev.hoot.api.widgets.Widgets;
import meteor.Main;
import meteor.api.items.Items;
import net.runelite.api.InventoryID;
import net.runelite.api.Item;
import net.runelite.api.ItemContainer;
import net.runelite.api.VarClientStr;
import net.runelite.api.widgets.Widget;
import net.runelite.api.widgets.WidgetID;
import net.runelite.api.widgets.WidgetInfo;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;
import java.util.function.Supplier;

public class Trade
{
	private static final Supplier<Widget> OUR_ITEMS = () -> Widgets.get(WidgetID.PLAYER_TRADE_SCREEN_GROUP_ID, 25);
	private static final Supplier<Widget> THEIR_ITEMS = () -> Widgets.get(WidgetID.PLAYER_TRADE_SCREEN_GROUP_ID, 28);
	private static final Supplier<Widget> INVENTORY = () -> Widgets.get(WidgetID.PLAYER_TRADE_INVENTORY_GROUP_ID, 0);
	private static final Supplier<Widget> ACCEPT_1 = () -> Widgets.get(WidgetID.PLAYER_TRADE_SCREEN_GROUP_ID, 10);
	private static final Supplier<Widget> ACCEPT_2 = () -> Widgets.get(WidgetInfo.SECOND_TRADING_WITH_ACCEPT_BUTTON);
	private static final Supplier<Widget> DECLINE_1 = () -> Widgets.get(WidgetID.PLAYER_TRADE_SCREEN_GROUP_ID, 13);
	private static final Supplier<Widget> DECLINE_2 = () -> Widgets.get(WidgetID.PLAYER_TRADE_CONFIRM_GROUP_ID, WidgetID.TradeScreen.SECOND_DECLINE_FUNC);

	public static boolean isOpen()
	{
		return isFirstScreenOpen() || isSecondScreenOpen();
	}

	public static boolean isSecondScreenOpen()
	{
		return Widgets.isVisible(ACCEPT_2.get());
	}

	public static boolean isFirstScreenOpen()
	{
		return Widgets.isVisible(ACCEPT_1.get());
	}

	public static void accept()
	{
		acceptFirstScreen();
		acceptSecondScreen();
	}

	public static void acceptFirstScreen()
	{
		Widget button = ACCEPT_1.get();
		if (Widgets.isVisible(button))
		{
			button.interact("Accept");
		}
	}

	public static void acceptSecondScreen()
	{
		Widget button = ACCEPT_2.get();
		if (Widgets.isVisible(button))
		{
			button.interact("Accept");
		}
	}

	public static void decline()
	{
		declineFirstScreen();
		declineSecondScreen();
	}

	public static void declineFirstScreen()
	{
		Widget button = DECLINE_1.get();
		if (Widgets.isVisible(button))
		{
			button.interact("Decline");
		}
	}

	public static void declineSecondScreen()
	{
		Widget button = DECLINE_2.get();
		if (Widgets.isVisible(button))
		{
			button.interact("Decline");
		}
	}

	public static void offer(Predicate<Item> filter, int quantity)
	{
		Item item = Trade.getInventory(filter).stream().findFirst().orElse(null);
		if (item == null)
		{
			return;
		}

		Item i = new Item(item.getId(), item.getQuantity());
		Items.offer(i, quantity);
	}

	public static void offer(int id, int quantity)
	{
		offer(x -> x.getId() == id, quantity);
	}

	public static void offer(String name, int quantity)
	{
		offer(x -> x.getName() != null && x.getName().equals(name), quantity);
	}

	/**
	 * Pass true as first parameter if you want to get the other side's items
	 */
	public static List<Item> getAll(boolean theirs, Predicate<Item> filter)
	{
		List<Item> items = new ArrayList<>();
		ItemContainer container = Game.getClient().getItemContainer(theirs ? InventoryID.TRADEOTHER : InventoryID.TRADE);
		if (container == null)
		{
			return items;
		}

		Item[] containerItems = container.getItems();
		for (Item item : containerItems) {
			if (item.getId() != -1 && item.getName() != null && !item.getName().equals("null")) {
				Widget containerWidget = theirs ? THEIR_ITEMS.get() : OUR_ITEMS.get();
				item.setWidgetId(item.calculateWidgetId(containerWidget));

				if (filter.test(item)) {
					items.add(item);
				}
			}
		}

		return items;
	}

	public static List<Item> getInventory(Predicate<Item> filter)
	{
		List<Item> items = new ArrayList<>();
		ItemContainer container = Game.getClient().getItemContainer(InventoryID.INVENTORY);
		if (container == null)
		{
			return items;
		}

		Item[] containerItems = container.getItems();
		for (Item item : containerItems) {
			if (item.getId() != -1 && item.getName() != null && !item.getName().equals("null")) {
				item.setWidgetId(item.calculateWidgetId(INVENTORY.get()));
				if (filter.test(item)) {
					items.add(item);
				}
			}
		}

		return items;
	}

	public static List<Item> getAll(boolean theirs)
	{
		return getAll(theirs, x -> true);
	}

	public static List<Item> getAll(boolean theirs, int... ids)
	{
		return getAll(theirs, Predicates.ids(ids));
	}

	public static List<Item> getAll(boolean theirs, String... names)
	{
		return getAll(theirs, Predicates.names(names));
	}

	public static Item getFirst(boolean theirs, Predicate<Item> filter)
	{
		return getAll(theirs, filter).stream().findFirst().orElse(null);
	}

	public static Item getFirst(boolean theirs, int... ids)
	{
		return getFirst(theirs, Predicates.ids(ids));
	}

	public static Item getFirst(boolean theirs, String... names)
	{
		return getFirst(theirs, Predicates.names(names));
	}

	public static boolean contains(boolean theirs, Predicate<Item> filter)
	{
		return getFirst(theirs, filter) != null;
	}

	public static boolean contains(boolean theirs, int... ids)
	{
		return contains(theirs, Predicates.ids(ids));
	}

	public static boolean contains(boolean theirs, String... names)
	{
		return contains(theirs, Predicates.names(names));
	}

	public static String getTradingPlayer()
	{
		return Vars.getVarcStr(VarClientStr.DUEL_OPPONENT_NAME);
	}
}

File: src/main/kotlin/meteor/api/items/Trade.kt
package meteor.api.items

import Main.client
import meteor.api.game.GameThread
import meteor.api.game.Vars
import meteor.api.widgets.Dialog
import meteor.api.widgets.Widgets
import net.runelite.api.InventoryID
import net.runelite.api.Item
import net.runelite.api.ItemContainer
import net.runelite.api.VarClientStr
import net.runelite.api.widgets.Widget
import net.runelite.api.widgets.WidgetID
import net.runelite.api.widgets.WidgetInfo
import java.util.ArrayList
import java.util.function.Predicate
import java.util.function.Supplier

object Trade {
    private val OUR_ITEMS = Supplier<Widget?> { Widgets[WidgetID.PLAYER_TRADE_SCREEN_GROUP_ID, 25] }
    private val THEIR_ITEMS = Supplier<Widget?> { Widgets[WidgetID.PLAYER_TRADE_SCREEN_GROUP_ID, 28] }
    private val INVENTORY = Supplier<Widget?> { Widgets[WidgetID.PLAYER_TRADE_INVENTORY_GROUP_ID, 0] }
    private val ACCEPT_1 = Supplier<Widget?> { Widgets[WidgetID.PLAYER_TRADE_SCREEN_GROUP_ID, 10] }
    private val ACCEPT_2 = Supplier<Widget?> { Widgets[WidgetInfo.SECOND_TRADING_WITH_ACCEPT_BUTTON] }
    private val DECLINE_1 = Supplier<Widget?> { Widgets[WidgetID.PLAYER_TRADE_SCREEN_GROUP_ID, 13] }
    private val DECLINE_2 = Supplier<Widget?> {
        Widgets.get(
            WidgetID.PLAYER_TRADE_CONFIRM_GROUP_ID,
            WidgetID.TradeScreen.SECOND_DECLINE_FUNC
        )
    }
    val isOpen: Boolean
        get() = isFirstScreenOpen || isSecondScreenOpen
    val isSecondScreenOpen: Boolean
        get() {
            val accept = ACCEPT_2.get()
            return accept != null && accept.isHidden
        }
    val isFirstScreenOpen: Boolean
        get() {
            val accept = ACCEPT_1.get()
            return accept != null && accept.isHidden
        }

    fun accept() {
        acceptFirstScreen()
        acceptSecondScreen()
    }

    fun acceptFirstScreen() {
        val button = ACCEPT_1.get()
        if (button != null && button.isHidden) {
            button.interact("Accept")
        }
    }

    fun acceptSecondScreen() {
        val button = ACCEPT_2.get()
        if (button != null && button.isHidden) {
            button.interact("Accept")
        }
    }

    fun decline() {
        declineFirstScreen()
        declineSecondScreen()
    }

    fun declineFirstScreen() {
        val button = DECLINE_1.get()
        if (button != null && button.isHidden) {
            button.interact("Decline")
        }
    }

    fun declineSecondScreen() {
        val button = DECLINE_2.get()
        if (button != null && button.isHidden) {
            button.interact("Decline")
        }
    }

    fun offer(filter: Predicate<Item>, quantity: Int) {
        val item = getInventory(filter).stream().findFirst().orElse(null) ?: return
        when (quantity) {
            1 -> item.interact("Offer")
            5 -> item.interact("Offer-5")
            10 -> item.interact("Offer-10")
            else -> if (quantity > item.quantity) {
                item.interact("Offer-All")
            } else {
                item.interact("Offer-X")
                Dialog.enterInput(quantity)
            }
        }
    }

    fun offer(id: Int, quantity: Int) {
        offer({ x: Item -> x.id == id }, quantity)
    }

    fun offer(name: String, quantity: Int) {
        offer({ x: Item -> x.name != null && x.name == name }, quantity)
    }

    /**
     * Pass true as first parameter if you want to get the other side's items
     */
    fun getAll(theirs: Boolean, filter: Predicate<Item>): List<Item?> {
        val items: MutableList<Item?> = ArrayList()
        val container: ItemContainer = client
            .getItemContainer(if (theirs) InventoryID.TRADEOTHER else InventoryID.TRADE)
            ?: return items
        Inventory.cacheItems(container)
        val containerItems = container.items
        var i = 0
        val containerItemsLength = containerItems.size
        while (i < containerItemsLength) {
            val item = containerItems[i]
            if (item.id != -1 && item.name != null && item.name != "null") {
                val containerWidget = if (theirs) THEIR_ITEMS.get() else OUR_ITEMS.get()
                item.widgetId = item.calculateWidgetId(containerWidget)
                item.slot = i
                if (filter.test(item)) {
                    items.add(item)
                }
            }
            i++
        }
        return items
    }

    fun getInventory(filter: Predicate<Item>): List<Item?> {
        val items: ArrayList<Item> = ArrayList()
        val container: ItemContainer = client.getItemContainer(InventoryID.INVENTORY)
            ?: return items
        Inventory.cacheItems(container)
        val containerItems = container.items
        var i = 0
        val containerItemsLength = containerItems.size
        while (i < containerItemsLength) {
            val item = containerItems[i]
            if (item.id != -1 && item.name != null && item.name != "null") {
                item.widgetId = item.calculateWidgetId(INVENTORY.get())
                item.slot = i
                if (filter.test(item)) {
                    items.add(item)
                }
            }
            i++
        }
        return items
    }

    fun getAll(theirs: Boolean): List<Item?> {
        return getAll(theirs) { x: Item? -> true }
    }

    fun getAll(theirs: Boolean, vararg ids: Int): List<Item?> {
        return getAll(theirs) { x: Item ->
            for (id in ids) {
                if (id == x.id) {
                    return@getAll true
                }
            }
            false
        }
    }

    fun getAll(theirs: Boolean, vararg names: String): List<Item?> {
        return getAll(theirs) { x: Item ->
            if (x.name == null) {
                return@getAll false
            }
            for (name in names) {
                if (name == x.name) {
                    return@getAll true
                }
            }
            false
        }
    }

    fun getFirst(theirs: Boolean, filter: Predicate<Item>): Item? {
        return getAll(theirs, filter).stream().findFirst().orElse(null)
    }

    fun getFirst(theirs: Boolean, vararg ids: Int): Item? {
        return getFirst(theirs) { x: Item ->
            for (id in ids) {
                if (id == x.id) {
                    return@getFirst true
                }
            }
            false
        }
    }

    fun getFirst(theirs: Boolean, vararg names: String): Item? {
        return getFirst(theirs) { x: Item ->
            if (x.name == null) {
                return@getFirst false
            }
            for (name in names) {
                if (name == x.name) {
                    return@getFirst true
                }
            }
            false
        }
    }

    fun contains(theirs: Boolean, filter: Predicate<Item>): Boolean {
        return getFirst(theirs, filter) != null
    }

    fun contains(theirs: Boolean, vararg ids: Int): Boolean {
        return contains(theirs) { x: Item ->
            for (id in ids) {
                if (id == x.id) {
                    return@contains true
                }
            }
            false
        }
    }

    fun contains(theirs: Boolean, vararg names: String): Boolean {
        return contains(theirs) { x: Item ->
            if (x.name == null) {
                return@contains false
            }
            for (name in names) {
                if (name == x.name) {
                    return@contains true
                }
            }
            false
        }
    }

    val tradingPlayer: String
        get() = Vars.getVarcStr(VarClientStr.DUEL_OPPONENT_NAME)
}